Prepare tutorial about binding Service Instances to lambda (#7302)
kyma-project/kyma: docs/serverless-v2/08-03-bind-service-instance.md

type: Tutorials
---

This tutorial shows how you can bind a sample instance of the Redis service to a lambda. After completing all steps, you will get the lambda with encoded Secrets to the service. You can use them for authentication when you connect to the service to implement custom business logic of your lambda.

To create the binding, you will use ServiceBinding and ServiceBindingUsage custom resources (CRs) managed by the Service Catalog.

>**NOTE:** To learn more about binding Service Instances to applications in Kyma, read [this](/components/service-catalog/#details-provisioning-and-binding) document.

## Prerequisites

This tutorial is based on an existing lambda. To create one, follow the [Create a lambda](#tutorials-create-a-lambda) tutorial.

## Steps

Follows these steps:

1. Export these variables:

    ```bash
    export NAME={LAMBDA_NAME}
    export NAMESPACE=serverless
    ```

    > **NOTE:** Lambda takes the name from the Function CR name. The ServiceInstance, ServiceBinding, and ServiceBindingUsage CRs can have different names, but for the purpose of this tutorial, all related resources share a common name defined under the **NAME** variable.

> 

Reviewer: ```
2. Provision an Addon CR with the Redis service:

cat <<EOF | kubectl apply -f  -
apiVersion: addons.kyma-project.io/v1alpha1
kind: AddonsConfiguration
metadata:
  name: $NAME
  namespace: $NAMESPACE
spec:
  reprocessRequest: 0
  repositories:
    - url: https://github.com/kyma-project/addons/releases/download/0.11.0/index-testing.yaml
EOF

3. Check if the Addon CR was created successfully. The CR phase should state `Phase: Ready`:

kubectl get addons $NAME -n $NAMESPACE -o=jsonpath="{.status.phase}"

```

> 2. Create a ServiceInstance CR. You will provision and use [Redis](https://redis.io/) with its `micro` plan:

Reviewer: ```suggestion
4. Create a ServiceInstance CR. You will use the provisioned [Redis](https://redis.io/) service with its `micro` plan:
```

Reviewer: Pls update the numbers below as well.

    ```yaml
    cat <<EOF | kubectl apply -f -
    apiVersion: servicecatalog.k8s.io/v1beta1
    kind: ServiceInstance
    metadata:
      name: $NAME
      namespace: $NAMESPACE
    spec:
      clusterServiceClassExternalName: redis
      clusterServicePlanExternalName: micro
      parameters:
        imagePullPolicy: Always
    EOF    

>     ```

Reviewer 2: Is the Redis available by default in Kyma? Maybe we should say that you need a ServiceInstance first and if not point to the proper tutorial?

Author: By default not. Also we haven't any tutorial for provisioning sample serviceClass in our documentation, so here we need to point to external docs, but we have a tutorial to bind in cluster Redis ServiceClass by Addon, but doc for that is outdated :( So my proposition is write as you wrote, user must firstly have Redis ServiceClass and for that we can add small yaml for provisioning Addon with Redis.

Reviewer: As discussed with @magicmatatjahu , there will be an additional step on adding yaml for provisioning an Addon with Redis.

Reviewer 2: Ok :smile:

    ```bash
    kubectl get servicebinding $NAME -n $NAMESPACE -o=jsonpath="{range .status.conditions[*]}{.type}{'\t'}{.status}{'\n'}{end}"
    ```

8. Create a ServiceBindingUsage CR:

    ```yaml
    cat <<EOF | kubectl apply -f -
    apiVersion: servicecatalog.kyma-project.io/v1alpha1
    kind: ServiceBindingUsage
    metadata:
      name: $NAME
      namespace: $NAMESPACE
    spec:
      serviceBindingRef:
        name: $NAME
      usedBy:
        kind: knative-service
        name: $NAME
      parameters:
        envPrefix:
          name: "REDIS_"
    EOF    
    ```

    - The **spec.serviceBindingRef** and **spec.usedBy** fields are required. **spec.serviceBindingRef** points to the Service Binding you have just created and **spec.usedBy** points to the lambda. More specifically, **spec.usedBy** refers to the name of the related KService CR (`name: $NAME`) and the cluster-specific [UsageKind CR](https://kyma-project.io/docs/components/service-catalog/#custom-resource-usage-kind) (`kind: knative-service`) that defines how Secrets should be injected to your lambda through the Service Binding.

>     > **NOTE:** Read [this](/components/service-catalog/#details-provisioning-and-binding) document to learn more about binding in Kyma.

Reviewer: Actually, we already have a similar note and the link at the beginning of the doc, so let's remove this one as it is a duplicate.

    - The **spec.parameters.envPrefix.name** field is optional. It adds a prefix to all environment variables injected by a given Secret from the Service Binding to the lambda. In our example, **envPrefix** is `REDIS_`, so all environmental variables will follow the `REDIS_{env}` naming pattern.

>     > **TIP:** It is considered good practice to use **envPrefix**. In some cases, lambda must use several instances of a given Service Class, so prefixes allow you to distinguish between instances and make sure that one Secret does not overwrite another one.

Reviewer: ```suggestion
    > **TIP:** It is considered good practice to use **envPrefix**. In some cases, a lambda must use several instances of a given Service Class. Prefixes allow you to distinguish between instances and make sure that one Secret does not overwrite another one.
```

> 

Reviewer: Could you adjust indents?

![image](https://user-images.githubusercontent.com/40655785/75853853-31113680-5def-11ea-8ba9-fca3e6fb9c31.png)

9. Check if the ServiceBindingUsage CR was created successfully. The last condition in the CR status should state `Ready True`:

    ```bash
    kubectl get servicebindingusage $NAME -n $NAMESPACE -o=jsonpath="{range .status.conditions[*]}{.type}{'\t'}{.status}{'\n'}{end}"
    ```

10. Retrieve and decode Secret details from the Service Binding:

    ```bash
    kubectl get secret $NAME -n $NAMESPACE -o go-template='{{range $k,$v := .data}}{{printf "%s: " $k}}{{if not $v}}{{$v}}{{else}}{{$v | base64decode}}{{end}}{{"\n"}}{{end}}'
    ```

    You should get a result similar to the following details:

    ```bash
    HOST: hb-redis-micro-0e965585-9699-443f-b987-38bc6af0e416-redis.serverless.svc.cluster.local
    PORT: 6379
    REDIS_PASSWORD: 1tvDcINZvp
    ```

>     > **NOTE:** If you added the **REDIS_** prefix for environmental variables in step 6, all variables will start with it. For example, the **PORT** variables will take the form of **REDIS_PORT**.

Reviewer: ```suggestion
    > **NOTE:** If you added the **REDIS_** prefix for environmental variables in step 6, all variables will start with it. For example, the **PORT** variable will take the form of **REDIS_PORT**.
```
